OCA/stock-logistics-warehouse
# [WMS][12.0] Add stock_reserve_rule

This module adds rules for advanced reservation / removal strategies.

Rules are applied on a location and its sub-locations.

A rule can exclude quants or locations based on configurable criteria,
and based on the selected quants, apply advanced removal strategies.

The rules have a sequence, which will be respected for the reservation.
So even without filter or advanced removal strategies, we can give a priority to
reserve in a location before another.

The advanced removal strategies are applied on top of the default one (fifo,
fefo, ...).

The included advanced removal strategies are:

* Default Removal Strategy: apply the default configured one (fifo, fefo, ...)
* Empty Bins: goods are removed from a bin only if the bin will be empty after
  the removal (favor smallest bins first, then apply the default removal
  strategy for equal quantities).
* Prefer Full Packaging: tries to remove full packaging (configured on the
  products) first, by largest to smallest package (default removal strategy is
  then applied for equal quantities).

Examples of scenario:

rules:

* location A: no filter, no advanced removal strategy
* location B: no filter, Empty Bins
* location C: no filter, no  advanced removal strategy

result:

* take what is available in location A
* then take in location B if available, only if bin(s) are emptied
* then take what is available in location C

The module is meant to be extensible, with a core mechanism on which new rules
and advanced removal strategies can be added.



It's part of the WMS addons to implement these specs:
https://docs.google.com/document/d/1mct6bFFWJqW01wGFcjc-uQNEjyCxvh6Y9TuFdRhe-b0/edit#heading=h.9ad56twdce31


Demo data has been created by the module.

Product: Funky Socks

Locations:

 * Stock / Zone A / Bin A1: 200 Funky socks
 * Stock / Zone B / Bin B1: 100 Funky socks
 * Stock / Zone C / Bin C1: 100 Funky socks

Reservation Rules, in the following order

 * Zone A must have full quantities
 * Zone B
 * Zone C

Delivery Orders:

* Origin: Outgoing shipment (reservation rules demo 1)
* Origin: Outgoing shipment (reservation rules demo 2)

Test:

* Activate Storage Locations and Multi-Warehouses
* You can open Inventory > Configuration > Stock Reservation Rules to see the rules
* Open Transfer: Outgoing shipment (reservation rules demo 1)
* Check availability: it has 150 units, as it will not empty Zone A, it will not take products there, it should take 100 in B and 50 in C (following the rules order)
* Unreserve this transfer (to test the second case)
* Open Transfer: Outgoing shipment (reservation rules demo 2)
* Check availability: it has 250 units, it can empty Zone A, it will take 200 in Bin A1 and 50 in Bin B1.
* If you want to explore further, you can add a custom domain to exclude rules (for instance, a product category will not use Zone B).
